### ISSUE TITLE
reader should cancel ongoing reads on timeout

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#11469]
As (the last remaining) part of https://github.com/scylladb/scylladb/issues/2927,
we should consider improving our timeout handling on top of https://github.com/scylladb/scylladb/commit/acf8da2bcef47168376e876e2fdc37fc26ebc82e (with the help of https://github.com/scylladb/seastar/commit/66b4c6acfdf0fa72d2397830f5306de7c41eaa64)
and cancel outstanding reads on timeout (e.g. using an abort_source and a timer), provided this doesn't degrade performance of the good path.

**Comment by bhalevy** [comment]
@xemul assigned to you provisionally.

**Comment by mykaul** [comment]
@bhalevy - how relevant is this today?

**Comment by bhalevy** [comment]
> [@bhalevy](https://github.com/bhalevy) - how relevant is this today?

It aged well.  I think we can close this issue.
WDYT @paszkow @denesb @xemul

**Comment by avikivity** [comment]
Why can we close it?

**Comment by bhalevy** [comment]
> Why can we close it?

IIRC we check the timeout when consuming mutations so readers may be blocked on I/O when the timeout expires but as soon as they check the deadline they should unwind and return an error.
That said, I can't point to the exact code that does that today.
@denesb?

**Comment by bhalevy** [comment]
> > Why can we close it?
> 
> IIRC we check the timeout when consuming mutations so readers may be blocked on I/O when the timeout expires but as soon as they check the deadline they should unwind and return an error. That said, I can't point to the exact code that does that today. [@denesb](https://github.com/denesb)?

Okay, so since 29a2788b2ed87f74cd5112a6abdd9b73d65d0359 the reader_permit::impl sets up a timer that sets an error on the permit if the timeout expires.  This is checked in e.g. in mx_sstable_mutation_reader fill_buffer loop, and as you said in https://github.com/scylladb/scylladb/issues/2927#issuecomment-1238325888, it's fine if the timeout is coarse.

Since we haven't seen any further problems with today's approach what's the point in keeping this issue open rather than closing it as "not planned"?

**Comment by avikivity** [comment]
I'm not saying it's wrong, I'm saying that saying "we can close it" without any explanation is insufficient.

We could consider canceling in-flight I/O, in case we don't already.

**Comment by bhalevy** [comment]
> I'm not saying it's wrong, I'm saying that saying "we can close it" without any explanation is insufficient.
> 
> We could consider canceling in-flight I/O, in case we don't already.

Today we don't cancel the read we're waiting for in fill_buffer, but as soon as we detect the timeout in check_abort, the reader returns an error and when it is destroyed along with the reader_permit, soon after, all in-flight read-aheads are canceled via the destroyed I/o intent member in the permit.

I think that's sufficient, hence it"s okay to close this issue as not planned.

**Comment by denesb** [comment]
I think we should also check for aborted read in the fill buffer loop of memtable and cache readers too. Those can also end up doing a lot of work on occasion.

**Comment by bhalevy** [comment]
> I think we should also check for aborted read in the fill buffer loop of memtable and cache readers too. Those can also end up doing a lot of work on occasion.

@denesb when reading from memory (memtable, cache)  think that checking abort in the fill_buffer loop is too intensive.
Maybe a better place would be in the consume_pausable helpers, before and/or after calling fill_buffer.

**Comment by denesb** [comment]
Why is polling a boolean once per filling an 8KB+ buffer too intensive?

**Comment by bhalevy** [comment]
> Why is polling a boolean once per filling an 8KB+ buffer too intensive?

Once per fill_buffer is fine of course.
I thought you meant in its inner loop that has mutation fragment resolution.

**Comment by denesb** [comment]
PR for timeout poll for cache and memtable: https://github.com/scylladb/scylladb/pull/28149
I also opened a new issue just for that, in case we decide to backport in the future: https://github.com/scylladb/scylladb/issues/28148
